### PR TITLE
FACT undeploy checkbox

### PR DESCRIPTION
**Pull request by TheChosenEvilOne**
This is the last checkbox from this:
https://github.com/OpenRA/OpenRA/pull/14209#issuecomment-338435111

Ill add the checkbox to TS little bit later
This makes Transforms PausableConditional

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Transforms.cs [hunk 1]
```diff
@@ -48,42 +48,47 @@ public class TransformsInfo : ITraitInfo
...
-	public class Transforms : IIssueOrder, IResolveOrder, IOrderVoice, IIssueDeployOrder
+	public class Transforms : ConditionalTrait<TransformsInfo>, IIssueOrder, IResolveOrder, IOrderVoice, IIssueDeployOrder
 	{
 		readonly Actor self;
-		readonly TransformsInfo info;
 		readonly BuildingInfo buildingInfo;
 		readonly string faction;
 
-		public Transforms(ActorInitializer init, TransformsInfo info)
+		public Transforms(ActorInitializer init, TransformsInfo info) : base(info)
```
Comment (MustaphaTR) [inline OpenRA.Mods.Common/Traits/Transforms.cs]: Style nit:
` : base(info)` goes to its own line.

```
		public Transforms(ActorInitializer init, TransformsInfo info)
			: base(info)
 		{
 			self = init.Self;
 			buildingInfo = self.World.Map.Rules.Actors[info.IntoActor].TraitInfoOrDefault<BuildingInfo>();
 			faction = init.Contains<FactionInit>() ? init.Get<FactionInit, string>() : self.Owner.Faction.InternalName;
 		}

File: mods/cnc/rules/structures.yaml [hunk 2]
```diff
@@ -22,6 +22,10 @@ FACT:
 		IntoActor: mcv
 		Offset: 1,1
 		Facing: 108
+		RequiresCondition: fact-undeploy
+	GrantConditionOnPrerequisite@GLOBALFACTUNDEPLOY:
+		Condition: fact-factundeploy
```
Comment (MustaphaTR) [inline mods/cnc/rules/structures.yaml]: I'm not sure about what corrent naming should be but this and 2 lines above don't match, so it wouldn't work and causing Travis to give error.

Comment (TheChosenEvilOne) [reply mods/cnc/rules/structures.yaml]: yeah I somehow messed that up

File: OpenRA.Mods.Common/Traits/Transforms.cs [hunk 3]
```diff
@@ -71,23 +72,22 @@ public string VoicePhraseForOrder(Actor self, Order order)
 
 		public bool CanDeploy()
 		{
-			if (IsTraitDisabled)
+			if (IsTraitPaused)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Transforms.cs]: You need to check `IsTraitDisabled` in here too!

File: OpenRA.Mods.Common/Traits/Transforms.cs [hunk 4]
```diff
@@ -71,23 +72,22 @@ public string VoicePhraseForOrder(Actor self, Order order)
...
 				return false;
+
 			return buildingInfo == null || self.World.CanPlaceBuilding(Info.IntoActor, buildingInfo, self.Location + Info.Offset, self);
 		}
 
 		public IEnumerable<IOrderTargeter> Orders
 		{
 			get
 			{
-				if (!IsTraitDisabled)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Transforms.cs]: This should stay.

File: OpenRA.Mods.Common/Traits/Transforms.cs [hunk 5]
```diff
@@ -133,7 +133,7 @@ public void DeployTransform(bool queued)
 
 		public void ResolveOrder(Actor self, Order order)
 		{
-			if (order.OrderString == "DeployTransform" && !IsTraitDisabled)
+			if (order.OrderString == "DeployTransform" && !IsTraitPaused)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Transforms.cs]: `!IsTraitPaused && !IsTraitDisabled`

**Comment by Arular101** [comment]
Do we also want to disable redeployable MCVs in the campaign? If yes, then we need to keep in mind a slight change should be made for the RA Allies-06a mission (in an other PR), explained [here](https://github.com/OpenRA/OpenRA/issues/12806#issuecomment-343433394) for [this](https://github.com/OpenRA/OpenRA/issues/12806#issuecomment-282298134) reason.

**Comment by GraionDilach** [comment]
Campaigns never had redeployable conyards.

File: mods/ts/rules/shared-structures.yaml [hunk 6]
```diff
@@ -32,6 +32,10 @@ GACNST:
 		Offset: 1,1
 		Facing: 96
 		DeployCursor: undeploy
+        RequiresCondition: factundeploy
```
Comment (MustaphaTR) [inline mods/ts/rules/shared-structures.yaml]: spaces --> tabs

Comment (TheChosenEvilOne) [reply mods/ts/rules/shared-structures.yaml]: yeah I noticed that myself

File: OpenRA.Mods.Common/Traits/Transforms.cs [hunk 7]
```diff
@@ -48,42 +48,48 @@ public class TransformsInfo : ITraitInfo
...
 
 		public string VoicePhraseForOrder(Actor self, Order order)
 		{
-			return (order.OrderString == "DeployTransform") ? info.Voice : null;
+			return (order.OrderString == "DeployTransform") ? Info.Voice : null;
 		}
 
 		public bool CanDeploy()
 		{
+			if (IsTraitPaused && IsTraitDisabled)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Transforms.cs]: Should be ||, no?

Comment (TheChosenEvilOne) [reply OpenRA.Mods.Common/Traits/Transforms.cs]: yeah

Comment (TheChosenEvilOne) [reply OpenRA.Mods.Common/Traits/Transforms.cs]: let me do that

**Review by GraionDilach — APPROVED** [review]
:+1:

**Comment by pchote** [comment]
The changes in each commit didn't match up with their commit messages (e.g. most of the mod rule changes also included further changes to Transforms.cs), so I reorganized and squashed them down.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/257a9024bfe928adb058a88c324e835e23e4b1bd

**Pull request merged**